rna/utils.py

```diff
@@ -0,0 +1,186 @@
+
+    assert sum(size) <= 1.0, 'The sum of the size for the train and test ' \
+                            'data must be must be equal to or below 1.0.'
+
+    indices_classes = indices_per_class(from_nhot_to_labels(y_nhot))
+
+    X_train, y_train, X_calibrate, y_calibrate, X_test, y_test = ([] for _ in range(6))
+
+    for indices_class in indices_classes:
+        indices = [i for i in range(len(indices_class))]
+        train_index, calibration_index, test_index = define_random_indices(indices, size)
+
+        X_for_class = X[indices_class]
+        y_nhot_for_class = y_nhot[indices_class]
+
+        X_train.extend(X_for_class[train_index])
+        y_train.extend(y_nhot_for_class[train_index])
+        X_calibrate.extend(X_for_class[calibration_index])
+        y_calibrate.extend(y_nhot_for_class[calibration_index])
+        X_test.extend(X_for_class[test_index])
+        y_test.extend(y_nhot_for_class[test_index])
+
+    print("The actual distribution (train, calibration, test) is ({}, {}, {})".format(
+        round(len(X_train)/X.shape[0], 3),
+        round(len(X_calibrate)/X.shape[0], 3),
+        round(len(X_test)/X.shape[0], 3))
+    )
+
+    X_train = np.array(X_train)
+    y_train = np.array(y_train)
+    X_calibrate = np.array(X_calibrate)
+    y_calibrate = np.array(y_calibrate)
+    X_test = np.array(X_test)
+    y_test = np.array(y_test)
+
+    return X_train, y_train, X_calibrate, y_calibrate, X_test, y_test
+
+
+def string2vec(list_of_strings, label_encoder):
+    """
+    Converts a list of strings of length N to an N x n_single_cell_types representation of 0s and 1s
+
+    :param list_of_strings: list of strings. Multiple cell types should be separated by and/or
+    :param label_encoder: LabelEncoder mapping strings to indices and vice versa
+    :return: n_mixures x n_celltypes matrix
+    """
+
+    target_classes = np.zeros((len(list_of_strings), len(single_cell_types)))
+    for i, list_item in enumerate(list_of_strings):
+        celltypes = list_item.split(' and/or ')
+        for celltype in celltypes:
+            target_classes[i, int(label_encoder.transform([celltype]))] = 1
+    return target_classes
+
+
+def from_nhot_to_labels(y_nhot):
+    """
+    Converts nhot encoded matrix into list with labels for unique rows.
+
+    :param y_nhot: nhot encoded matrix
+    :return: list of length y_nhot.shape[0]
+    """
+
+    unique_labels = np.flip(np.unique(y_nhot, axis=0), axis=1)
+    if np.array_equal(np.sum(unique_labels, axis=1), np.ones(y_nhot.shape[1])):
+        y = np.argmax(y_nhot, axis=1)
+    else:
+        # assumes that the nhot encoded matrix first row consists of zero's
+        # and a 1 is added each row starting from the right.
+        y = []
+        for i in range(unique_labels.shape[0]):
+            y += [i] * np.where(np.all(y_nhot == unique_labels[i], axis=1))[0].shape[0]
+
+    return y
+
+# TODO: Make this function faster
+def replace_labels(y_nhot):
```

Comment:
ik snap niet helemaal waar deze voor is?

Reply:
Dit was een tijdelijke oplossing, dus niet meer nodig